lib/connextclient/ConnextClient.ts

```diff
@@ -0,0 +1,699 @@
+ * @param errorTimeout optional maximum duration of time to wait for the preimage
+ */
+const waitForPreimageByHash = (
+  client: SwapClient,
+  expectedHash: string,
+  errorTimeout = 30000,
+): Promise<string> => {
+  // create an observable that emits values when a preimage
+  // event is triggered on the client
+  const preimage$: Observable<ProvidePreimageEvent> = fromEvent(client, 'preimage');
+  const expectedPreimageReceived$ = preimage$.pipe(
+    // filter out events that do not match our expected hash
+    filter(preimageEvent => preimageEvent.rHash === expectedHash),
+    // map the ProvidePreimageEvent object to a string as the consumer is
+    // only interested in the value of the preimage
+    pluck('preimage'),
+    // complete the observable and clean up all the listeners
+    // once we receive 1 event that matches our conditions
+    take(1),
+    // emit an error if the observable does not emit any values
+    // for the specified duration
+    timeout(errorTimeout),
+  );
+  // convert the observable to a promise that resolves on complete
+  // and rejects on error
+  return expectedPreimageReceived$.toPromise();
+};
+
+/**
+ * A class representing a client to interact with connext.
+ */
+class ConnextClient extends SwapClient {
+  public readonly type = SwapClientType.Connext;
+  public readonly finalLock = 200;
+  public address?: string;
+  /** A map of currency symbols to token addresses. */
+  public tokenAddresses = new Map<string, string>();
+  private expectedIncomingTransfers = new Map<string, ExpectedIncomingTransfer>();
+  private port: number;
+  private host: string;
+  private webhookport: number;
+  private webhookhost: string;
+  private unitConverter: UnitConverter;
+  private seed: string | undefined;
+
+  /**
+   * Creates a connext client.
+   */
+  constructor({
+    config,
+    logger,
+    unitConverter,
+    currencyInstances,
+  }: {
+    unitConverter: UnitConverter;
+    config: ConnextClientConfig;
+    currencyInstances: CurrencyInstance[],
+    logger: Logger;
+  }) {
+    super(logger, config.disable);
+
+    this.port = config.port;
+    this.host = config.host;
+    this.webhookhost = config.webhookhost;
+    this.webhookport = config.webhookport;
+    this.unitConverter = unitConverter;
+    this.setTokenAddresses(currencyInstances);
+  }
+
+  public get minutesPerBlock() {
+    return 0.25; // 15 seconds per block target
+  }
+
+  public get label() {
+    return 'Connext';
+  }
+
+  public initSpecific = async () => {
+    // listen for incoming transfers
+    this.on('transferReceived', (transferReceivedRequest: TransferReceivedEvent) => {
+      const {
+        tokenAddress,
+        units,
+        timelock,
+        rHash,
+      } = transferReceivedRequest;
+      const expectedIncomingTransfer = this.expectedIncomingTransfers.get(rHash);
+      if (!expectedIncomingTransfer) {
+        this.logger.error(`received unexpected incoming transfer with rHash ${rHash}`);
```

Comment:
Is `error` maybe too severe here? I'm not sure how unexpected it would be to get an incoming transfer we are not expecting.